core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -203,32 +203,46 @@ ManifestReader<F> scanMetrics(ScanMetrics newScanMetrics) {
   }
 
   CloseableIterable<ManifestEntry<F>> entries() {
-    if ((rowFilter != null && rowFilter != Expressions.alwaysTrue())
-        || (partFilter != null && partFilter != Expressions.alwaysTrue())
-        || (partitionSet != null)) {
+    return entries(false /* all entries */);
+  }
+
+  private CloseableIterable<ManifestEntry<F>> entries(boolean onlyLive) {
+    if (hasRowFilter() || hasPartitionFilter() || partitionSet != null) {
       Evaluator evaluator = evaluator();
       InclusiveMetricsEvaluator metricsEvaluator = metricsEvaluator();
 
       // ensure stats columns are present for metrics evaluation
       boolean requireStatsProjection = requireStatsProjection(rowFilter, columns);
       Collection<String> projectColumns =
           requireStatsProjection ? withStatsColumns(columns) : columns;
+      CloseableIterable<ManifestEntry<F>> entries =
+          open(projection(fileSchema, fileProjection, projectColumns, caseSensitive));
 
       return CloseableIterable.filter(
           content == FileType.DATA_FILES
               ? scanMetrics.skippedDataFiles()
               : scanMetrics.skippedDeleteFiles(),
-          open(projection(fileSchema, fileProjection, projectColumns, caseSensitive)),
+          onlyLive ? filterLiveEntries(entries) : entries,
```

Comment:
The deleted entries must be discarded prior to evaluating stats primarily to have correct skipped counts. It should also help the performance.